File: pharmion-backend/Pharmion.Services/Services/ChronicDiseaseService.cs
using MapsterMapper;
using Microsoft.EntityFrameworkCore;
using Pharmion.Model.Exceptions;
using Pharmion.Model.Requests;
using Pharmion.Model.Responses;
using Pharmion.Model.SearchObjects;
using Pharmion.Services.Database;
using Pharmion.Services.Database.Entities;
using Pharmion.Services.Interfaces;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Pharmion.Services.Services
{
    public class ChronicDiseaseService : BaseCRUDService<ChronicDiseaseResponse, ChronicDiseaseSearchObject, ChronicDisease, ChronicDiseaseUpsertRequest, ChronicDiseaseUpsertRequest>, IChronicDiseaseService
    {
        private readonly PharmionDbContext _context;
        public ChronicDiseaseService(PharmionDbContext context, IMapper mapper) : base(context, mapper)
        {
            _context = context;
        }

        protected override IQueryable<ChronicDisease> ApplyFilter(IQueryable<ChronicDisease> query, ChronicDiseaseSearchObject search)
        {
            if (!string.IsNullOrEmpty(search.Code))
            {
                query = query.Where(cd => cd.Code.Contains(search.Code));
            }
            if (!string.IsNullOrEmpty(search.Name))
            {
                query = query.Where(cd => cd.Name.Contains(search.Name));
            }
            if (search.IsActive.HasValue)
            {
                query = query.Where(cd => cd.IsActive == search.IsActive.Value);
            }

            if (!string.IsNullOrEmpty(search.FTS))
            {
                query = query.Where(cd => cd.Code.Contains(search.FTS)
                                       || cd.Name.Contains(search.FTS)
                                       || cd.Description.Contains(search.FTS));
            }

            return query;
        }

        protected override async Task BeforeInsert(ChronicDisease entity, ChronicDiseaseUpsertRequest request)
        {
            if (await _context.ChronicDiseases.AnyAsync(cd => cd.Code == request.Code))
                throw new UserException("Chronic disease with this code already exists.");

            if (await _context.ChronicDiseases.AnyAsync(cd => cd.Name == request.Name))
                throw new UserException("Chronic disease with this name already exists.");

            entity.CreatedAt = DateTime.UtcNow;
        }

        protected override async Task BeforeUpdate(ChronicDisease entity, ChronicDiseaseUpsertRequest request)
        { 
            if (await _context.ChronicDiseases.AnyAsync(cd => cd.Code == request.Code && cd.Id != entity.Id))
                throw new UserException("Chronic disease with this code already exists.");

            if (await _context.ChronicDiseases.AnyAsync(cd => cd.Name == request.Name && cd.Id != entity.Id))
                throw new UserException("Chronic disease with this name already exists.");

            entity.UpdatedAt = DateTime.UtcNow;
        }

        protected override void MapUpdateToEntity(ChronicDisease entity, ChronicDiseaseUpsertRequest request)
        {
            base.MapUpdateToEntity(entity, request);
        }

        protected override async Task BeforeDelete(ChronicDisease entity)
        {
            var hasPatients = await _context.PatientChronicDiseases.AnyAsync(pcd => pcd.ChronicDiseaseId == entity.Id);

            if (hasPatients)
                throw new UserException("Cannot delete chronic disease: it is assigned to one or more patients.");
        }

        public override async Task<PagedResult<ChronicDiseaseResponse>> GetAsync(ChronicDiseaseSearchObject search)
        {
            var baseQuery = _context.Set<ChronicDisease>().AsQueryable();

            baseQuery = ApplyFilter(baseQuery, search);

            int? totalCount = null;
            if (search.IncludeTotalCount)
            {
                totalCount = await baseQuery.CountAsync();
            }

            var entities = await baseQuery.ToListAsync();

            var responseList = entities.Select(cd => new ChronicDiseaseResponse
            {
                Id = cd.Id,
                Code = cd.Code,
                Name = cd.Name,
                Description = cd.Description,
                IsActive = cd.IsActive,
                CreatedAt = cd.CreatedAt,
                UpdatedAt = cd.UpdatedAt
            }).ToList();

            if (!string.IsNullOrWhiteSpace(search.OrderBy))
            {
                var orderBy = search.OrderBy.ToLower();
                var descending = orderBy.StartsWith("-");
                if (descending)
                {
                    orderBy = orderBy.Substring(1);
                }

                responseList = orderBy switch
                {
                    "code" => descending
                        ? responseList.OrderByDescending(x => x.Code).ToList()
                        : responseList.OrderBy(x => x.Code).ToList(),
                    "name" => descending
                        ? responseList.OrderByDescending(x => x.Name).ToList()
                        : responseList.OrderBy(x => x.Name).ToList(),
                    "isactive" => descending
                        ? responseList.OrderByDescending(x => x.IsActive).ToList()
                        : responseList.OrderBy(x => x.IsActive).ToList(),
                    "createdat" => descending
                        ? responseList.OrderByDescending(x => x.CreatedAt).ToList()
                        : responseList.OrderBy(x => x.CreatedAt).ToList(),
                    "updatedat" => descending
                        ? responseList.OrderByDescending(x => x.UpdatedAt).ToList()
                        : responseList.OrderBy(x => x.UpdatedAt).ToList(),
                    _ => responseList
                };
            }

            if (!search.RetrieveAll && search.Page.HasValue && search.PageSize.HasValue)
            {
                responseList = responseList
                    .Skip(search.Page.Value * search.PageSize.Value)
                    .Take(search.PageSize.Value)
                    .ToList();
            }

            return new PagedResult<ChronicDiseaseResponse>
            {
                Items = responseList,
                TotalCount = totalCount
            };
        }
    }
}

File: pharmion-frontend/pharmion_desktop/lib/services/chronic_disease_service.dart
import 'api_service.dart';

class ChronicDiseaseModel {
  final int id;
  final String code;
  final String name;
  final String description;
  final bool isActive;
  final DateTime createdAt;
  final DateTime? updatedAt;

  const ChronicDiseaseModel({
    required this.id,
    required this.code,
    required this.name,
    required this.description,
    required this.isActive,
    required this.createdAt,
    this.updatedAt,
  });

  factory ChronicDiseaseModel.fromJson(Map<String, dynamic> json) =>
      ChronicDiseaseModel(
        id: json['id'] as int? ?? 0,
        code: json['code'] as String? ?? '',
        name: json['name'] as String? ?? '',
        description: json['description'] as String? ?? '',
        isActive: json['isActive'] as bool? ?? true,
        createdAt: DateTime.tryParse(json['createdAt'] ?? '') ?? DateTime(2000),
        updatedAt: json['updatedAt'] != null
            ? DateTime.tryParse(json['updatedAt'])
            : null,
      );
}

class PagedChronicDiseases {
  final List<ChronicDiseaseModel> items;
  final int totalCount;
  const PagedChronicDiseases({required this.items, required this.totalCount});
}

class ChronicDiseaseService {
  static Future<PagedChronicDiseases> getDiseases({
    int page = 0,
    int pageSize = 10,
    String? name,
    bool? isActive,
  }) async {
    final params = StringBuffer('ChronicDisease?includeTotalCount=true');
    params.write('&page=$page&pageSize=$pageSize');
    if (name != null && name.isNotEmpty) params.write('&name=$name');
    if (isActive != null) params.write('&isActive=$isActive');

    final data =
        await ApiService.get(params.toString()) as Map<String, dynamic>;
    final items = ((data['items'] as List?) ?? [])
        .map((e) => ChronicDiseaseModel.fromJson(e as Map<String, dynamic>))
        .toList();

    return PagedChronicDiseases(
      items: items,
      totalCount: data['totalCount'] as int? ?? items.length,
    );
  }

  static Future<ChronicDiseaseModel> create(Map<String, dynamic> body) async {
    final data =
        await ApiService.post('ChronicDisease', body) as Map<String, dynamic>;
    return ChronicDiseaseModel.fromJson(data);
  }

  static Future<ChronicDiseaseModel> update(
    int id,
    Map<String, dynamic> body,
  ) async {
    final data =
        await ApiService.put('ChronicDisease/$id', body)
            as Map<String, dynamic>;
    return ChronicDiseaseModel.fromJson(data);
  }

  static Future<void> delete(int id) async {
    await ApiService.delete('ChronicDisease/$id');
  }
}
